### PR TITLE
Reshape should support reshape to zero shapes

### PR DESCRIPTION
**Pull request by ilyachur**
### Details:
 - Try to fix reshape to zero shapes

### Tickets:
 - 45277

### CODE REVIEW

File: ngraph/core/src/op/reshape.cpp [hunk 1]
```diff
@@ -49,9 +49,8 @@ namespace reshapeop
                                 vector<Dimension>& output_shape)
     {
         if (reshape_pattern == std::vector<Dimension>{0} && !reshape_node->get_special_zero())
-        { // legacy check introduced by PR #1206
-            reshape_pattern = std::vector<Dimension>{};
-            output_shape = {};
+        {
+            output_shape = {0};
```
Comment (slyalin) [inline ngraph/core/src/op/reshape.cpp:53]: Please use original value of `reshape_patter` to keep possible dims tracking.

File: ngraph/core/src/op/reshape.cpp [hunk 2]
```diff
@@ -49,9 +49,8 @@ namespace reshapeop
                                 vector<Dimension>& output_shape)
     {
         if (reshape_pattern == std::vector<Dimension>{0} && !reshape_node->get_special_zero())
```
Comment (jane-intel) [inline ngraph/core/src/op/reshape.cpp:51]: simply delete this part of code

Comment (ilyachur) [reply ngraph/core/src/op/reshape.cpp]: Thanks! Fixed

File: ngraph/core/src/op/reshape.cpp (hunk 1 above)
Comment (ilyachur) [reply ngraph/core/src/op/reshape.cpp]: Fixed

**Review by jane-intel — APPROVED** [review]

**Comment by ilyachur** [comment]
@slyalin @jane-intel Please have a look one more time to this PR. This PR contains backward incompatible changes. I changed Myriad tests in order to avoid degradations.

**Review by jane-intel — APPROVED** [review]

**Comment by ilyachur** [comment]
Myriad tests `DynamicToStaticTopKPropagationConcatReshape` are failing with issue: `The 'K' input must be a scalar.

File: ngraph/core/src/op/reshape.cpp [hunk 3]
```diff
@@ -217,10 +211,17 @@ void op::v1::Reshape::validate_and_infer_types()
...
-                          "Pattern shape must have rank 1, got ",
+                          shape_pattern_shape.rank().compatible(1) ||
+                              (shape_pattern_shape.rank().is_static() &&
+                               shape_pattern_shape.rank().get_length() == 0),
+                          "Pattern shape must have rank 1 or be empty, got ",
                           shape_pattern_shape.rank(),
                           ".");
+    if (shape_pattern_shape.rank().is_static() && shape_pattern_shape.rank().get_length() == 0)
+    {
+        set_output_type(0, get_input_element_type(0), input_pshape);
```
Comment (jane-intel) [inline ngraph/core/src/op/reshape.cpp:222]: I'm a bit confused. Reshape([1], []) should result in [], but as I understand the code -- it would be [1].

Comment (ilyachur) [reply ngraph/core/src/op/reshape.cpp]: Correct, fixed

File: ngraph/core/src/op/reshape.cpp [hunk 4]
```diff
@@ -35,158 +35,12 @@ namespace reshapeop
...
         }
     }
-
-    void calculate_output_shape(const ngraph::op::v1::Reshape* reshape_node,
-                                vector<Dimension>& reshape_pattern,
-                                const int64_t& minus_one_idx,
-                                const PartialShape& input_pshape,
-                                vector<Dimension>& output_shape)
-    {
-        if (reshape_pattern == std::vector<Dimension>{0} && !reshape_node->get_special_zero())
```
Comment (ilyachur) [inline ngraph/core/src/op/reshape.cpp:51]: I removed this If

**Comment by ilyachur** [comment]
@jane-intel @slyalin Please have a look

**Review by jane-intel — CHANGES_REQUESTED** [review]
I dislike the idea that we don't change calculate_output_shape method.

**Comment by ilyachur** [comment]
> I dislike the idea that we don't change calculate_output_shape method. Why do we only change the code in validate and infer types?

Why do we need to change the method which works as expected? In validate and infer type we only changed the input arguments for `calculate_output_shape`.

File: ngraph/test/backend/reshape.in.cpp [hunk 5]
```diff
@@ -331,6 +331,37 @@ NGRAPH_TEST(${BACKEND_NAME}, builder_reshape_3D_to_scalar)
     test_case.run();
 }
 
+NGRAPH_TEST(${BACKEND_NAME}, builder_reshape_1d_to_same_shape)
```
Comment (ilyachur) [inline ngraph/test/backend/reshape.in.cpp:334]: @jane-intel Added more tests

File: ngraph/core/src/op/reshape.cpp [hunk 6]
```diff
@@ -250,13 +109,22 @@ void op::v1::Reshape::validate_and_infer_types()
             }
             reshape_pattern.emplace_back(lower_bound[i], upper_bound[i]);
         }
+        // For scalar case reshape_patter should be empty but scalar reshape pattern should be empty
+        // or equal to 1
+        if (output_rank.is_static() && output_rank.get_length() == 0 && !lower_bound.empty())
+        {
+            reshape_pattern.clear();
```
Comment (jane-intel) [inline ngraph/core/src/op/reshape.cpp:116]: why do we need to clear it up?

**Review by jane-intel — APPROVED** [review]

**Pull request merged**